### PR TITLE
Carrier - Optimise when updating Carrier

### LINKED ISSUE #36263: Optimize the carrier update, avoid generating a new instance on each save, handle duplication ONLY when the carrier is already assigned to an order
**US**
Optimize the carrier update, avoid generating a new instance on each save, handle duplication ONLY when the carrier is already assigned to an order

Impact on module ??

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR aim to optimise the way of update Carrier. In fact, we need to duplicate and set `deleted = true` to the old one, if the carrier has order linked. If not, we use update the old one.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI 🟢
| UI Tests          | ~
| Fixed issue or discussion?     | Fixes #36263 
| Related PRs       | ~
| Sponsor company   | PrestaShop SA

### CODE REVIEW

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 1]
```diff
@@ -69,73 +79,133 @@ public function add(Carrier $carrier): CarrierId
...
+
+    public function carrierHasOrders(CarrierId $carrierId): bool
+    {
+        $qb = $this->connection->createQueryBuilder();
+
+        $count = $qb->select('COUNT(*)')
+            ->from($this->dbPrefix . 'order_carrier', 'oc')
+            ->where('oc.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue())
+            ->groupBy('oc.id_carrier')
```
Comment (mflasquin) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]: I think the group by is not necessary, no ?

Comment (boherm) [reply src/Adapter/Carrier/Repository/CarrierRepository.php]: Oh yes, you're right! Thanks! ;) It's fixed

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 2]
```diff
@@ -69,73 +79,132 @@ public function add(Carrier $carrier): CarrierId
...
+    }
+
+    /**
+     * Create a new version of the carrier, or return the carrier as is if it don't have order linked.
+     *
+     * @param Carrier $carrier
+     *
+     * @return Carrier
+     */
+    public function createNewVersion(Carrier $carrier): Carrier
```
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]:
```suggestion
    public function getEditableOrNewVersion(Carrier $carrier): Carrier
```

File: src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php [hunk 3]
```diff
@@ -56,76 +58,77 @@ public function __construct(
      */
     public function handle(EditCarrierCommand $command): CarrierId
     {
+        // Get new version of carrier if needed
         $carrier = $this->carrierRepository->get($command->getCarrierId());
+        $newCarrier = $this->carrierRepository->createNewVersion($carrier);
```
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php]:
```suggestion
        $carrier = $this->carrierRepository->createNewVersion($command->getCarrierId());
```

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 4]
```diff
@@ -69,73 +79,132 @@ public function add(Carrier $carrier): CarrierId
...
+    {
+        $carrierId = new CarrierId((int) $carrier->id);
+
+        // If the carrier don't have orders linked, we can return it as is
+        if (!$this->carrierHasOrders($carrierId)) {
+            return $carrier;
```
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]:
```suggestion
            return $this->get($carrierId);
```

File: src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php [hunk 5]
```diff
@@ -141,6 +144,12 @@ public function handle(EditCarrierCommand $command): CarrierId
             }
         }
 
+        $this->carrierRepository->update(
```
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/EditCarrierHandler.php]: Move to line 131

File: src/Adapter/Carrier/CommandHandler/SetCarrierRangesHandler.php [hunk 6]
```diff
@@ -51,8 +51,10 @@ public function __construct(
      */
     public function handle(SetCarrierRangesCommand $command): CarrierId
     {
-        // Get new version of carrier
-        $newCarrier = $this->carrierRepository->updateInNewVersion($command->getCarrierId(), new Carrier());
+        // Get new version of carrier if needed
+        $carrier = $this->carrierRepository->get($command->getCarrierId());
+        $newCarrier = $this->carrierRepository->createNewVersion($carrier);
```
Comment (jolelievre) [inline src/Adapter/Carrier/CommandHandler/SetCarrierRangesHandler.php]:
```suggestion
        $carrier = $this->carrierRepository->createNewVersion($command->getCarrierId());
```

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 7]
```diff
@@ -69,73 +79,132 @@ public function add(Carrier $carrier): CarrierId
             CannotAddCarrierException::class
         );
 
-        return new CarrierId($carrierId);
+        return new CarrierId((int) $carrierId);
     }
 
-    public function updateInNewVersion(CarrierId $carrierId, Carrier $carrier): Carrier
+    public function update(Carrier $carrier, ShopConstraint $shopConstraint, int $errorCode): void
```
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]: I think you can simply use `updateObjectModel` and no need for the shop constraint parameter either

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 8]
```diff
@@ -69,73 +79,132 @@ public function add(Carrier $carrier): CarrierId
...
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     *
+     * @return ShopId[]
+     */
+    public function getAssociatedShopIds(CarrierId $carrierId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
```
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]: You can use `getObjectModelAssociatedShopIds`

File: src/Adapter/Carrier/Repository/CarrierRepository.php [hunk 9]
```diff
@@ -69,73 +79,132 @@ public function add(Carrier $carrier): CarrierId
...
         $this->updateObjectModel($newCarrier, CannotUpdateCarrierException::class);
+        $newCarrier->setGroups($carrier->getAssociatedGroupIds());
 
-        $newCarrier->setGroups($oldCarrier->getAssociatedGroupIds());
-
+        // Return the new duplicated carrier
         return $newCarrier;
     }
+
+    public function carrierHasOrders(CarrierId $carrierId): bool
```
Comment (jolelievre) [inline src/Adapter/Carrier/Repository/CarrierRepository.php]:
```suggestion
    protected function carrierHasOrders(CarrierId $carrierId): bool
```

File: tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature [hunk 10]
```diff
@@ -73,10 +89,18 @@ Feature: Carrier management
       | shippingMethod   | weight                             |
       | taxRuleGroup     | US-AL Rate (4%)                    |
       | rangeBehavior    | disabled                           |
+    When I update order "bo_order1" Tracking number to "TEST1234" and Carrier to "carrier1"
+    Then order "bo_order1" has Tracking number "TEST1234"
     When I edit carrier "carrier1" called "newCarrier1" with specified properties:
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature]:
```suggestion
    When I edit carrier "carrier1" with specified properties:
```
This endpoint will check automatically that the returned ID is the same

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Carrier/carrier_management.feature]:
```suggestion
    When I edit carrier "carrier1" with specified properties I get a new carrier called "newCarrier1":
```
This one checks that the IDs ARE different

**Review by matks — APPROVED** [review]
It's great you covered that with tests 👍

**Review by matthieu-rolland — APPROVED** [review]

**Pull request merged**